geopandas/io/sql.py

```diff
@@ -47,17 +47,26 @@ def read_postgis(sql, con, geom_col='geom', crs=None, hex_encoded=True,
     if geom_col not in df:
         raise ValueError("Query missing geometry column '{}'".format(geom_col))
 
-    def load_geom(x):
-        if isinstance(x, bytes):
-            return shapely.wkb.loads(x, hex=hex_encoded)
-        else:
-            return shapely.wkb.loads(str(x), hex=hex_encoded)
-    geoms = df[geom_col].apply(load_geom)
-    df[geom_col] = geoms
+    load_geom_bytes = shapely.wkb.loads
 
-    if crs is None:
-        if len(geoms) > 0:
-            srid = shapely.geos.lgeos.GEOSGetSRID(geoms[0]._geom)
+    def load_geom_buffer(x):
+        return shapely.wkb.loads(str(x))
+
+    def load_geom_str(x):
+        return shapely.wkb.loads(str(x), hex=hex_encoded)
+    geoms = df[geom_col].dropna()
+    if not geoms.empty:
+        import sys
+        df[geom_col] = geoms = geoms.apply(
+            (
+                load_geom_buffer if isinstance(geoms.iat[0], buffer)
+                else load_geom_str
+            ) if sys.version_info.major < 3
+            else load_geom_bytes if isinstance(geoms.iat[0], bytes)
+            else load_geom_str
```

Comment:
Rewritten in new commits.
Functional style `if` expressions tell me the evaluation is purely logical, and I won't have to consider complex assignments (does every branch assign to the same variable?) or worry about impossible errors (are any of those repeated variable assignments mistyped?), but I agree they look awkward in Python.
Why do conditional branching outside `load_geom`*? Databases define column types, so rather than check every single value in a column, I thought it better to check once and treat all values as a single type.